File: app/src/main/java/com/ywwynm/everythingdone/activities/EverythingDoneBaseActivity.java
package com.ywwynm.everythingdone.activities;

import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.SparseArray;
import android.view.View;

import com.ywwynm.everythingdone.permission.PermissionCallback;
import com.ywwynm.everythingdone.utils.DeviceUtil;
import com.ywwynm.everythingdone.utils.LocaleUtil;

/**
 * Created by ywwynm on 2015/6/4.
 * A base Activity class to reduce same codes in different subclasses.
 */
public abstract class EverythingDoneBaseActivity extends AppCompatActivity {

    public static final String TAG = "EverythingDoneBaseActivity";

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Call this before setContentView so that NavigationView can update its language correctly.
        LocaleUtil.changeLanguage();

        beforeSetContentView();

        setContentView(getLayoutResource());

        beforeInit();
        init();
    }

    @SuppressWarnings("unchecked")
    protected final <T extends View> T f(@IdRes int id) {
        return (T) findViewById(id);
    }

    @SuppressWarnings("unchecked")
    protected final <T extends View> T f(View v, @IdRes int id) {
        return (T) v.findViewById(id);
    }

    protected void beforeSetContentView() {}

    protected abstract @LayoutRes int getLayoutResource();

    protected void beforeInit() {}

    protected void init() {
        initMembers();
        findViews();
        initUI();
        setActionbar();
        setEvents();
    }

    protected abstract void initMembers();

    protected abstract void findViews();

    protected abstract void initUI();

    protected abstract void setActionbar();

    protected abstract void setEvents();

    private SparseArray<PermissionCallback> mCallbacks;

    public void doWithPermissionChecked(
            @NonNull PermissionCallback permissionCallback, int requestCode, String... permissions) {
        if (DeviceUtil.hasMarshmallowApi()) {
            if (mCallbacks == null) {
                mCallbacks = new SparseArray<>();
            }
            for (String permission : permissions) {
                int pg = ContextCompat.checkSelfPermission(this, permission);
                if (pg != PackageManager.PERMISSION_GRANTED) {
                    mCallbacks.put(requestCode, permissionCallback);
                    ActivityCompat.requestPermissions(this, permissions, requestCode);
                    return;
                }
            }
        }

        permissionCallback.onGranted();
    }

    @Override
    public void onRequestPermissionsResult(
            int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        PermissionCallback callback = mCallbacks.get(requestCode);
        if (callback != null) {
            for (int grantResult : grantResults) {
                if (grantResult != PackageManager.PERMISSION_GRANTED) {
                    callback.onDenied();
                    return;
                }
            }
            callback.onGranted();
        }
    }

}

File: app/src/main/java/com/ywwynm/everythingdone/activities/EverythingDoneBaseActivity.kt
package com.ywwynm.everythingdone.activities

import android.content.Context
import android.content.pm.PackageManager
import android.os.Bundle
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.appcompat.app.AppCompatActivity
import android.util.SparseArray
import android.view.View

import com.ywwynm.everythingdone.permission.PermissionCallback
import com.ywwynm.everythingdone.utils.LocaleUtil

/**
 * Created by ywwynm on 2015/6/4.
 * Translated to Kotlin by ywwynm and Claude Opus 4.7 on 2026/5/20.
 * A base Activity class to reduce same codes in different subclasses.
 */
abstract class EverythingDoneBaseActivity : AppCompatActivity() {

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(LocaleUtil.attachBaseContext(newBase))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Keep AppCompat's per-app locale state in sync before inflating views.
        LocaleUtil.changeLanguage()

        beforeSetContentView()

        setContentView(getLayoutResource())

        beforeInit()
        init()
    }

    @Suppress("UNCHECKED_CAST")
    protected fun <T : View?> f(@IdRes id: Int): T {
        return findViewById<View>(id) as T
    }

    @Suppress("UNCHECKED_CAST")
    protected fun <T : View?> f(v: View, @IdRes id: Int): T {
        return v.findViewById<View>(id) as T
    }

    protected open fun beforeSetContentView() {}

    @LayoutRes
    protected abstract fun getLayoutResource(): Int

    protected open fun beforeInit() {}

    protected open fun init() {
        initMembers()
        findViews()
        initUI()
        setActionbar()
        setEvents()
    }

    protected abstract fun initMembers()

    protected abstract fun findViews()

    protected abstract fun initUI()

    protected abstract fun setActionbar()

    protected abstract fun setEvents()

    private var mCallbacks: SparseArray<PermissionCallback>? = null

    open fun doWithPermissionChecked(
        permissionCallback: PermissionCallback, requestCode: Int, vararg permissions: String?
    ) {
        if (mCallbacks == null) {
            mCallbacks = SparseArray()
        }
        for (permission in permissions) {
            val pg = ContextCompat.checkSelfPermission(this, permission!!)
            if (pg != PackageManager.PERMISSION_GRANTED) {
                mCallbacks!!.put(requestCode, permissionCallback)
                @Suppress("UNCHECKED_CAST")
                ActivityCompat.requestPermissions(this, permissions as Array<String>, requestCode)
                return
            }
        }

        permissionCallback.onGranted()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (mCallbacks == null) return
        val callback: PermissionCallback = mCallbacks!!.get(requestCode) ?: return
        for (grantResult in grantResults) {
            if (grantResult != PackageManager.PERMISSION_GRANTED) {
                callback.onDenied()
                return
            }
        }
        callback.onGranted()
    }

    companion object {
        const val TAG: String = "EverythingDoneBaseActivity"
    }
}
